### PR TITLE
feat(rust): move trezor-client build script to separate crate

### LINKED ISSUE #3153: Rust: `trezor-client`'s build script fails when `protoc` binary is not present
**Describe the bug**
After the [`0.1.1` release](https://crates.io/crates/trezor-client/versions), [`trezor-client`](https://github.com/trezor/trezor-firmware/tree/b4825f408a5140af43ffcc1744e0df3c5fdf07be/rust/trezor-client) fails building when `protoc` binary is not present due to the addition of the build script.
This makes our CI [workflows fail on `ethers-rs`](https://github.com/gakonst/ethers-rs/actions/runs/5561045481/jobs/10159775784?pr=2510) when building with support for Trezor.

Given that the Rust protobuf bindings are platform agnostic, can their generation be included in the main makefile pipeline like with python's? This would remove the `protoc` dependency and compilation step from its users.

I'd be willing to work on this, if this is desired.

### PR DESCRIPTION
**Pull request by DaniPopes**
Fixes #3153

- Create `trezor-client-build`, which generates the protobuf bindings
  - This is run with `rust/trezor-client/scripts/build_protos` or `make protobuf`, with an optional `--check` flag
- Drive-by: remove Cargo.lock (see [Cargo FAQ](https://doc.rust-lang.org/cargo/faq.html#why-do-binaries-have-cargolock-in-version-control-but-not-libraries))
- Drive-by: build docs.rs documentation with `--cfg docsrs` and add `#[feature(doc_auto_cfg)]` for displaying features alongside items (see [docs.rs](https://docs.rs/about/builds#cross-compiling))

### CODE REVIEW

**Comment by DaniPopes** [comment]
Hello @prusnak @matejcik can I at least get a CI run on this :smile:

**Comment by prusnak** [comment]
Can you please resolve the conflict in `rust/trezor-client/Cargo.lock`? 

I will then try to run the CI.

**Comment by DaniPopes** [comment]
Sure, done

**Comment by DaniPopes** [comment]
Rebuilt bindings, checks should pass now

**Review by mmilata — COMMENTED** [review]
Changes look good to me, would like @matejcik to have brief look as well.

CI: https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/pipelines/1020026907

File: rust/trezor-client/src/messages/mod.rs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+/// This macro provides the TrezorMessage trait for a protobuf message.
+macro_rules! trezor_message_impl {
+	($($struct:ident => $mtype:expr),+ $(,)?) => {$(
+		impl TrezorMessage for $struct {
+			const MESSAGE_TYPE: MessageType = $mtype;
+		}
+	)+};
+}
```
[the remark this answers is not in the document]
Comment (DaniPopes) [reply rust/trezor-client/src/messages/mod.rs:24]: Sure, looks like rustfmt avoids this macro

**Comment by matejcik** [comment]
is there a reason to complicate the crate, as opposed to install and run protobuf-codegen from the shell script?

> Drive-by: remove Cargo.lock (see [Cargo FAQ](https://doc.rust-lang.org/cargo/faq.html#why-do-binaries-have-cargolock-in-version-control-but-not-libraries))

please revert, the guidance has changed recently and this has the potential to mess up our CI for no reason

**Comment by DaniPopes** [comment]
> is there a reason to complicate the crate, as opposed to install and run protobuf-codegen from the shell script?

Wdym by complicate the crate? The new -build crate does the same thing as the previous `build.rs` just not as a build script that runs on users machine

> please revert, the guidance has changed recently and this has the potential to mess up our CI for no reason

Yup, done

**Comment by matejcik** [comment]
> Wdym by complicate the crate? The new -build crate does the same thing as the previous build.rs just not as a build script that runs on users machine

yeah but now you introduced workspaces (with strange version declarations that took me quite some time to decode) and you still need to run the crate via a magic invocation of `cargo run` from the shell script.
which, AFAICT, could be replaced by a `protoc --rust_out path/to/generated common/protob/*.proto` if the protobuf_codegen plugin is present, because there doesn't seem to be any customization?

**Comment by matejcik** [comment]
> * we need both the codegen dependency (`protobuf-codegen` crate) and the runtime dependency (`protobuf` crate) to be the same version
> * I think `--rust_out` is unstable at the protoc versions you are using in this repo

oof
okay, yeah, let's do that like this.

i'm probably gonna replace the whole thing with, well, something else, because for whatever reason the "default" language bindings for protobuf are invariably bad. but no reason to do that here.

**Review by matejcik — APPROVED** [review]

**Comment by prusnak** [comment]
Squashed and merged. Thanks!

**Pull request merged**